src/useImageDimensions.ts

```diff
@@ -0,0 +1,61 @@
+import {useEffect, useState} from 'react'
+import {Image, ImageRequireSource} from 'react-native'
+
+export interface URISource {
+  uri: string
+}
+
+/**
+ * @param source either a remote URL or a local file resource.
+ * @returns original image dimensions (width, height and aspect ratio).
+ */
+function useImageDimensions(source: ImageRequireSource | URISource) {
+  const [[dimensions, error], setState] = useState<
+    [{width: number; height: number}?, Error?]
+  >([])
```

Comment:
I think that this could be a bit easier to read (took me a while to figure out what was happening 😅) if you use two separate states:

```
const [dimensions, setDimensions] = useState<{width: number; height: number} | null>(null)
const [error, setError] = useState<Error | null>(null)
```

Reply:
With two different states, there will be more rows of code, as well as more calls of setState, which possibly will not invoke rerender multiple times, but still, some code inside the box will execute once again. Also, I'm not sure, but possibly 2 different states will take more RAM.
It still looks ok with one state with an array inside I think.
p.s.: maybe it is strange, I'm thinking a lot of performance because now I'm busy with research to develop applications for smart-TVs.

Reply:
Less code != more readable. I think it's much clearar with more than one `setState`.

> I'm thinking a lot of performance because now I'm busy with research to develop applications for smart-TVs.

I think it's good to have performance in mind, but this seems like premature optimisation to me.

We also have no benchmarks to show that this would indeed take less RAM, e.g. you're allocating arrays here which you wouldn't have to do in the other case...

Reply:
Well...  "На вкус и цвет товарища нет". Which means "Every man to his own taste" ;)